### PR TITLE
Add type defintions for actions-on-google

### PR DESCRIPTION
**Pull request by joelhegg**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
> actions-on-google makes thorough use of jsdoc, so tsd-jsdoc gave me a better starting point.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

### CODE REVIEW

File: types/actions-on-google/index.d.ts
Comment (plantain-00) [inline types/actions-on-google/index.d.ts]: Looks OK, except that I'm not sure whether the Apache 2.0 License declaration conflicts with the repository's MIT license.

**Review by plantain-00 — APPROVED** [review]

File: types/actions-on-google/index.d.ts
Comment (joelhegg) [reply types/actions-on-google/index.d.ts]: This isn't the first time this project has run into this. Some examples of existing type definitions tagged with an Apache license: applepayjs, bootstrap-switch, core-js, date.format.js, es6-collections, google-apps-script, highland, winrt.

Comment (plantain-00) [reply types/actions-on-google/index.d.ts]: Take `@types/core-js` as example, I `npm install` it, I can find a `LICENSE` file of `MIT License`, and `"license": "MIT"` in `package.json` file, but the code has an Apache 2.0 License declaration, it confuses me what its license is.
@andy-ms

Comment (Thomas-P) [reply types/actions-on-google/index.d.ts]: Hi, 

the Apache 2.0 License is the license on which the project is published. But the Apache 2.0 License allows you to publish own software under your license, when you pointed out that the original software is published under Apache 2.0, which is already present in the header of each of the file.

So @joelhegg is it possible, to add the MIT License to the root of the project and in the package.json file? 
This would be the license of your own work and this commit.

**Review by ghost — CHANGES_REQUESTED** [review]
We should probably get legal review before publishing something owned by an outside entity (either Google or @pilwon) unless we're sure we have permission to.

**Comment by joelhegg** [comment]
The reason I kept the license headers is because 80% of what's in here is derived from Google's original jsdoc. Apache and MIT are both permissive licenses and should be compatible, but Apache still expects acknowledgement (https://opensource.stackexchange.com/questions/1713/adding-feature-to-an-apache-v2-package).

@pilwon's changes are largely a stripped down version of Google's original code with Typescript typing added. He actually offered to contribute this back to Google in actions-on-google/actions-on-google-nodejs#2, but they weren't interested in maintaining it. IANAL, but perhaps this falls under the "Submission of Contributions" section of the Apache v2 license.

I can check with Google in actions-on-google/actions-on-google-nodejs#2 to see if they want to give their signoff on this PR, but I don't think it's required by the Apache v2 license, and I wouldn't count on them responding. I'm afraid if we make this a blocker for this PR then it won't go anywhere.

**Comment by andy-ms** [comment]
Would it be acceptable to remove the license but leave an acknowledgement?

File: types/actions-on-google/index.d.ts
Comment (RyanCavanaugh) [reply types/actions-on-google/index.d.ts]: Those should not have been merged. Different licenses are a very big problem for us because we automated tasks that depend on the license being an appropriate one in order for the tool's behavior to be legal.

I'm not super concerned about Apache (seems like it's OK but I haven't cleared it with our legal team) but unless there's a very, very good reason to introduce more problems here, I can't merge this.

Comment (joelhegg) [reply types/actions-on-google/index.d.ts]: So it sounds like the options are:
1. @RyanCavanaugh to clear the usage of the Apache license within type definitions with his legal folks.
1. Remove the license headers as suggested by @andy-ms and add a NOTICE similar to https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/google-apps-script/NOTICE
1. Use dts-gen to generate a clean type definition, but lose several thousand lines of documentation.
1. Abandon the PR.

Please let me know how you'd like to proceed.

**Comment by RyanCavanaugh** [comment]
I will contact our legal team to see if we can allow Apache contributions. This comes up reasonably often and it seems like it shouldn't be too difficult.

File: types/actions-on-google/index.d.ts
Comment (joelhegg) [reply types/actions-on-google/index.d.ts]: @plantain-00 @Thomas-P @andy-ms @RyanCavanaugh Based on the following response from Google in actions-on-google/actions-on-google-nodejs#2 I removed the license headers and added a NOTICE. Please let me know if this addresses your concerns.

> @joelhegg from what I've seen the pull request that you have submitted doesn't contain any files authored by us, so we have no claims on them. You can strip the headers and change the license, although you should probably check with @pilwon as it looks like your work is based off what he's done. For the jsdoc comments, as you mentioned you can attribute them to our docs released under CC, adding a NOTICE similar to the one you point to in your comments.
>
> As @Thomas-P commented on the DefinitelyTyped pull request, in a repo like @pilwon's fork where you have some files authored by us and some not, you can change the license for the files you authored, you'd only need to keep our headers in the file authored by us, a copy of the apache license in the repo (as per point 4a of the license), notices for files authored by us that you have changed.

**Comment by weswigham** [comment]
@RyanCavanaugh Any update?

**Comment by andy-ms** [comment]
Waiting to hear back from @DanielRosenwasser before merging this

**Comment by joelhegg** [comment]
@andy-ms Let's try this again. Google signed off on this in actions-on-google/actions-on-google-nodejs#2:

> @joelhegg from what I've seen the pull request that you have submitted doesn't contain any files authored by us, so we have no claims on them. You can strip the headers and change the license, although you should probably check with @pilwon as it looks like your work is based off what he's done. For the jsdoc comments, as you mentioned you can attribute them to our docs released under CC, adding a NOTICE similar to the one you point to in your comments.

Based on that I removed the license headers and added a NOTICE.

I never heard back from @pilwon, so I wasn't comfortable basing this on his code. I deleted the existing files and generated clean ones using tsd-jsdoc (the commits show my work step by step). actions-on-google's jsdoc is pretty thorough, so this got me most of the way there.

At this point all licensing concerns should be resolved, so this should be set to merge.

**Comment by RyanCavanaugh** [comment]
This PR has been open and unchanged 5 days without signoff or complaint. This will be merged by a maintainer soon if there are no objections.

**Comment by RyanCavanaugh** [comment]
@andy-ms - Thanks for your review of this PR! Can you please look at the new code and update your review status if appropriate?

**Comment by andy-ms** [comment]
@DanielRosenwasser

**Comment by joelhegg** [comment]
@andy-ms Can you elaborate on what your concerns are at this point?

**Comment by DanielRosenwasser** [comment]
Basically we're waiting on a legal team to let us know if there are concerns about publishing Apache-2.0-licensed `.d.ts` files to `@types` (even though we've been publishing other packages for months).

**Comment by joelhegg** [comment]
@DanielRosenwasser See my comment from 10/20. There is no Apache-2.0 licensed code here anymore. I deleted everything and auto-generated it fresh from the original Javascript library using tsd-jsdoc. Even then, Google confirmed nothing in the original code really needed to remain under the Apache-2.0 license.

That's cool if your legal team confirms DefinitelyTyped's compatibility with Apache-2.0, but it should be orthogonal to this PR at this point.

**Comment by DanielRosenwasser** [comment]
Thanks a ton for being patient! It should be in now.

**Pull request merged**
